File: app/src/main/java/ru/magflayer/spectrum/presentation/pages/main/camera/ColorCameraFragment.java
package ru.magflayer.spectrum.presentation.pages.main.camera;

import android.graphics.SurfaceTexture;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v7.graphics.Palette;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.view.OrientationEventListener;
import android.view.TextureView;
import android.view.View;
import android.view.ViewGroup;
import android.view.WindowManager;
import android.view.animation.AccelerateInterpolator;
import android.view.animation.Animation;
import android.view.animation.TranslateAnimation;
import android.widget.TextView;
import android.widget.Toast;
import android.widget.ToggleButton;

import com.arellomobile.mvp.presenter.InjectPresenter;

import java.util.List;

import butterknife.BindDimen;
import butterknife.BindView;
import butterknife.OnClick;
import lombok.AllArgsConstructor;
import lombok.Getter;
import ru.magflayer.spectrum.R;
import ru.magflayer.spectrum.domain.injection.InjectorManager;
import ru.magflayer.spectrum.presentation.common.android.BaseFragment;
import ru.magflayer.spectrum.presentation.common.android.layout.Layout;
import ru.magflayer.spectrum.presentation.common.android.widget.ColorDetailsWidget;
import ru.magflayer.spectrum.presentation.common.android.widget.PointView;
import ru.magflayer.spectrum.presentation.common.model.SurfaceInfo;
import ru.magflayer.spectrum.presentation.common.utils.ViewUtils;

@Layout(R.layout.fragment_color_camera)
public class ColorCameraFragment extends BaseFragment implements TextureView.SurfaceTextureListener, ColorCameraView {

    private final static int ROTATION_INTERVAL = 5;

    @InjectPresenter
    ColorCameraPresenter presenter;

    @BindView(R.id.camera)
    TextureView cameraView;
    @BindView(R.id.right_menu)
    ViewGroup buttonsMenuView;
    @BindView(R.id.left_menu)
    ViewGroup infoMenuView;

    @BindView(R.id.menu)
    View menuButton;
    @BindView(R.id.color_recycler)
    RecyclerView colorRecycler;
    @BindView(R.id.toggle_mode)
    ToggleButton toggleView;
    @BindView(R.id.point_detector)
    PointView pointView;
    @BindView(R.id.color_details)
    ColorDetailsWidget colorDetailsWidget;
    @BindView(R.id.message)
    TextView messageView;

    @BindDimen(R.dimen.color_list_width)
    protected int menuSize;

    private ColorCameraAdapter adapter;
    private OrientationEventListener orientationEventListener;
    private Orientation currentOrientation;

    public static ColorCameraFragment newInstance() {
        return new ColorCameraFragment();
    }

    @Override
    protected void inject() {
        InjectorManager.getAppComponent().inject(this);
    }

    @Override
    public void onViewCreated(@NonNull final View view, @Nullable final Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);

        presenter.setRouter(getRouter());
        orientationEventListener = new OrientationEventListener(getContext()) {
            @Override
            public void onOrientationChanged(int orientation) {
                if (isLandscape(orientation) && currentOrientation != Orientation.LANDSCAPE) {
                    currentOrientation = Orientation.LANDSCAPE;
                    setOrientation(currentOrientation);
                } else if (isPortrait(orientation) && currentOrientation != Orientation.PORTRAIT) {
                    currentOrientation = Orientation.PORTRAIT;
                    setOrientation(currentOrientation);
                }
            }
        };

        if (orientationEventListener.canDetectOrientation()) {
            orientationEventListener.enable();
        } else {
            orientationEventListener.disable();
        }

        updateMode(toggleView.isChecked());
    }

    @Override
    public void onDestroyView() {
        orientationEventListener.disable();
        super.onDestroyView();
    }

    @Override
    public void onActivityCreated(final Bundle savedInstanceState) {
        super.onActivityCreated(savedInstanceState);
        getActivity().getWindow().addFlags(WindowManager.LayoutParams.FLAG_KEEP_SCREEN_ON);

        colorRecycler.setLayoutManager(new LinearLayoutManager(getContext(), LinearLayoutManager.HORIZONTAL, false));
        adapter = new ColorCameraAdapter(getContext());
        colorRecycler.setAdapter(adapter);
        cameraView.setSurfaceTextureListener(this);
        colorRecycler.setVisibility(toggleView.isChecked() ? View.GONE : View.VISIBLE);
        toggleView.setOnCheckedChangeListener((buttonView, isChecked) -> updateMode(isChecked));
    }

    @Override
    public void onSurfaceTextureAvailable(final SurfaceTexture surface, final int width, final int height) {
        logger.debug("onSurfaceTextureAvailable");
        presenter.handleSurfaceAvailable(surface);
    }

    @Override
    public void onSurfaceTextureSizeChanged(final SurfaceTexture surface, final int width, final int height) {
        //do nothing
    }

    @Override
    public boolean onSurfaceTextureDestroyed(final SurfaceTexture surface) {
        logger.debug("onSurfaceTextureDestroyed");
        presenter.handleSurfaceDestroyed();
        return true;
    }

    @Override
    public void onSurfaceTextureUpdated(final SurfaceTexture surface) {
        presenter.updateSurface(toggleView.isChecked() ? SurfaceInfo.Type.SINGLE : SurfaceInfo.Type.MULTIPLE);
    }

    @Override
    public void showPictureSavedToast() {
        Toast.makeText(getContext(), R.string.camera_image_saved, Toast.LENGTH_SHORT).show();
    }

    @Override
    public void showColors(final List<Palette.Swatch> colors) {
        adapter.setData(colors);
    }

    @Override
    public void showColorDetails(final int mainColor, final int titleColor) {
        colorDetailsWidget.setColor(mainColor);
        pointView.setAimColor(titleColor);
    }

    @Override
    public void showColorName(final String name) {
        colorDetailsWidget.setColorName(name);
    }

    @Override
    public void showErrorMessage() {
        messageView.setVisibility(View.VISIBLE);
    }

    @Override
    public void hideErrorMessage() {
        messageView.setVisibility(View.GONE);
    }

    @Override
    public void showCrosshair() {
        pointView.setVisibility(View.VISIBLE);
    }

    @Override
    public void hideCrosshair() {
        pointView.setVisibility(View.GONE);
    }

    @Override
    public void showPanels() {
        showTopMenu();
        showBottomMenu();
    }

    @OnClick(R.id.camera)
    protected void onFocusClick() {
        presenter.handleFocusClicked();
    }

    @OnClick(R.id.menu)
    protected void onMenuClick() {
        presenter.openHistory();
    }

    @OnClick(R.id.save)
    protected void onSaveClick() {
        presenter.handleSaveClicked();
    }

    private void showTopMenu() {
        if (buttonsMenuView.getVisibility() != View.VISIBLE) {
            buttonsMenuView.setVisibility(View.VISIBLE);
            buttonsMenuView.startAnimation(inFromTopAnimation());
        }
    }

    private void showBottomMenu() {
        if (infoMenuView.getVisibility() != View.VISIBLE) {
            infoMenuView.setVisibility(View.VISIBLE);
            infoMenuView.startAnimation(inFromBottomAnimation());
        }
    }

    private Animation inFromTopAnimation() {
        Animation inFromTop = new TranslateAnimation(
                Animation.RELATIVE_TO_PARENT, 0.0f,
                Animation.RELATIVE_TO_PARENT, 0.0f,
                Animation.RELATIVE_TO_PARENT, +1.0f,
                Animation.RELATIVE_TO_PARENT, 0.0f);
        inFromTop.setDuration(800);
        inFromTop.setInterpolator(new AccelerateInterpolator());
        return inFromTop;
    }

    private Animation inFromBottomAnimation() {
        Animation inFromBottom = new TranslateAnimation(
                Animation.RELATIVE_TO_PARENT, 0.0f,
                Animation.RELATIVE_TO_PARENT, 0.0f,
                Animation.RELATIVE_TO_PARENT, -1.0f,
                Animation.RELATIVE_TO_PARENT, 0.0f);
        inFromBottom.setDuration(800);
        inFromBottom.setInterpolator(new AccelerateInterpolator());
        return inFromBottom;
    }

    private void setOrientation(final Orientation orientation) {
        logger.debug("Orientation changed: {}", orientation);
        ViewUtils.rotateView(menuButton, orientation.getDegree());
        ViewUtils.rotateView(toggleView, orientation.getDegree());
        colorDetailsWidget.rotate(orientation.getDegree());
    }

    private boolean isPortrait(final int orientation) {
        return orientation <= ROTATION_INTERVAL || orientation >= 360 - ROTATION_INTERVAL // [350 : 10]
                || (orientation <= 180 + ROTATION_INTERVAL && orientation >= 180 - ROTATION_INTERVAL); //[170 : 190]
    }

    private boolean isLandscape(final int orientation) {
        return (orientation <= 90 + ROTATION_INTERVAL && orientation >= 90 - ROTATION_INTERVAL) // [100 : 80]
                || (orientation <= 270 + ROTATION_INTERVAL && orientation >= 270 - ROTATION_INTERVAL); // [280 : 260]
    }

    private void updateMode(final boolean single) {
        if (single) {
            ViewUtils.changeViewVisibility(true, colorDetailsWidget, pointView);
            ViewUtils.changeViewVisibility(false, colorRecycler);
        } else {
            ViewUtils.changeViewVisibility(false, colorDetailsWidget, pointView);
            ViewUtils.changeViewVisibility(true, colorRecycler);
        }
    }

    @Getter
    @AllArgsConstructor
    private enum Orientation {
        PORTRAIT(0),
        LANDSCAPE(90);

        private int degree;
    }
}

File: app/src/main/java/ru/magflayer/spectrum/presentation/pages/main/camera/ColorCameraFragment.kt
package ru.magflayer.spectrum.presentation.pages.main.camera

import android.annotation.SuppressLint
import android.os.Bundle
import android.view.*
import android.view.animation.AccelerateInterpolator
import android.view.animation.Animation
import android.view.animation.TranslateAnimation
import android.widget.Toast
import androidx.core.view.GestureDetectorCompat
import androidx.palette.graphics.Palette
import androidx.recyclerview.widget.GridLayoutManager
import by.kirich1409.viewbindingdelegate.viewBinding
import dagger.hilt.EntryPoint
import dagger.hilt.InstallIn
import dagger.hilt.android.EntryPointAccessors
import dagger.hilt.android.components.ActivityComponent
import moxy.presenter.InjectPresenter
import moxy.presenter.ProvidePresenter
import ru.magflayer.spectrum.R
import ru.magflayer.spectrum.databinding.FragmentColorCameraBinding
import ru.magflayer.spectrum.presentation.common.android.BaseFragment
import ru.magflayer.spectrum.presentation.common.extension.hide
import ru.magflayer.spectrum.presentation.common.extension.rotate
import ru.magflayer.spectrum.presentation.common.extension.show
import ru.magflayer.spectrum.presentation.pages.main.camera.holder.ColorAnalyzer
import ru.magflayer.spectrum.presentation.pages.main.camera.holder.ColorCameraHolder

class ColorCameraFragment : BaseFragment(R.layout.fragment_color_camera), ColorCameraView {

    companion object {

        private const val ZOOM_VISIBLE_DELAY = 1000L //ms
        private const val COLOR_SPAN_COUNT = 2
        private const val MENU_ANIMATION_DURATION = 800L

        fun newInstance(): ColorCameraFragment {
            return ColorCameraFragment()
        }
    }

    @InjectPresenter
    lateinit var presenter: ColorCameraPresenter

    private val viewBinding by viewBinding(FragmentColorCameraBinding::bind)
    private val cameraHolder by lazy { ColorCameraHolder(requireContext()) }
    private val gestureDetector by lazy {
        GestureDetectorCompat(
            requireContext(),
            gestureDetectorListener
        )
    }
    private val gestureDetectorListener = object : GestureDetector.SimpleOnGestureListener() {

        override fun onScroll(
            e1: MotionEvent?,
            e2: MotionEvent?,
            distanceX: Float,
            distanceY: Float
        ): Boolean {
            presenter.handleCameraZoom(distanceX, distanceY)
            return super.onScroll(e1, e2, distanceX, distanceY)
        }
    }
    private val orientationEventListener by lazy {
        object : OrientationEventListener(requireContext()) {
            override fun onOrientationChanged(orientationDegree: Int) {
                presenter.handleOrientationChanged(orientationDegree)
            }
        }
    }
    private val colorAnalyzer by lazy {
        ColorAnalyzer().apply {
            analyzeResultListener = { presenter.handleAnalyzeImage(it) }
        }
    }

    private lateinit var adapter: ColorCameraAdapter

    @EntryPoint
    @InstallIn(ActivityComponent::class)
    interface ColorCameraEntryPoint {
        fun colorCameraPresenter(): ColorCameraPresenter
    }

    @ProvidePresenter
    fun providePresenter(): ColorCameraPresenter {
        return EntryPointAccessors.fromActivity(
            requireActivity(),
            ColorCameraEntryPoint::class.java
        ).colorCameraPresenter()
    }

    @SuppressLint("ClickableViewAccessibility")
    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        requireActivity().window.addFlags(WindowManager.LayoutParams.FLAG_KEEP_SCREEN_ON)

        if (orientationEventListener.canDetectOrientation()) {
            orientationEventListener.enable()
        } else {
            orientationEventListener.disable()
        }

        viewBinding.toggleMode.setOnCheckedChangeListener { _, checked ->
            presenter.handleColorModeChanged(checked)
        }
        viewBinding.menu.setOnClickListener { presenter.handleMenuClicked() }
        viewBinding.save.setOnClickListener {
            cameraHolder.takePicture(
                onSuccess = { presenter.handlePictureCaptureSucceed(it) },
                onError = { presenter.handlePictureCaptureFailed(it) }
            )
        }
        viewBinding.flash.setOnClickListener { presenter.handleFlashClick(viewBinding.flash.isChecked) }

        adapter = ColorCameraAdapter(requireContext())
        val layoutManager = GridLayoutManager(
            requireContext(),
            COLOR_SPAN_COUNT,
            GridLayoutManager.HORIZONTAL,
            false
        )
        viewBinding.colorRecycler.layoutManager = layoutManager
        viewBinding.colorRecycler.adapter = adapter

        viewBinding.cameraPreview.setOnClickListener { presenter.handleFocusClicked() }
        viewBinding.cameraPreview.setOnTouchListener { _, motionEvent ->
            gestureDetector.onTouchEvent(motionEvent)
        }

        cameraHolder.startCamera(
            viewLifecycleOwner,
            viewBinding.cameraPreview,
            colorAnalyzer
        ) { cameraInfo ->
            presenter.handleCameraInitialized(cameraInfo)
        }
    }

    override fun onDestroyView() {
        super.onDestroyView()
        orientationEventListener.disable()
        requireActivity().window.clearFlags(WindowManager.LayoutParams.FLAG_KEEP_SCREEN_ON)
    }

    override fun showPictureSavedToast() {
        Toast.makeText(context, R.string.camera_image_saved, Toast.LENGTH_SHORT).show()
    }

    override fun showColors(colors: List<Palette.Swatch>) {
        adapter.setData(colors)
    }

    override fun showColorDetails(mainColor: Int, titleColor: Int) {
        viewBinding.colorDetails.setColor(mainColor)
        viewBinding.pointDetector.setAimColor(titleColor)
    }

    override fun showColorName(name: String) {
        viewBinding.colorDetails.setColorName(name)
    }

    override fun showErrorMessage() {
        viewBinding.message.show()
    }

    override fun hideErrorMessage() {
        viewBinding.message.hide()
    }

    override fun showCrosshair() {
        viewBinding.pointDetector.show()
    }

    override fun hideCrosshair() {
        viewBinding.pointDetector.hide()
    }

    override fun showPanels() {
        showTopMenu()
        showBottomMenu()
    }

    override fun showFlash() {
        viewBinding.flash.show()
    }

    override fun hideFlash() {
        viewBinding.flash.hide()
    }

    override fun enableFlash() {
        cameraHolder.enabledFlash()
    }

    override fun disableFlash() {
        cameraHolder.disableFlash()
    }

    override fun showZoom(zoom: Float, maxZoom: Int) = with(viewBinding) {
        zoomSeek.max = maxZoom
        zoomSeek.progress = zoom.toInt()

        if (zoomSeek.visibility != View.VISIBLE) {
            zoomSeek.show()
            zoomSeek.postDelayed({ zoomSeek.hide() }, ZOOM_VISIBLE_DELAY)
        }

        cameraHolder.setZoomRatio(zoom)
    }

    override fun showSingleColorMode() = with(viewBinding) {
        colorDetails.show()
        pointDetector.show()
        colorRecycler.hide()
        colorAnalyzer.analyzerType = ColorAnalyzer.Type.CENTER
    }

    override fun showMultipleColorMode() = with(viewBinding) {
        colorDetails.hide()
        pointDetector.hide()
        colorRecycler.show()
        colorAnalyzer.analyzerType = ColorAnalyzer.Type.SWATCHES
    }

    override fun updateViewOrientation(orientation: CameraOrientation) = with(viewBinding) {
        menu.rotate(orientation.degree)
        toggleMode.rotate(orientation.degree)
        flash.rotate(orientation.degree)
        colorDetails.rotate(orientation.degree)
        zoomContainer.rotate((orientation.degree - 90).rem(360))
        viewBinding.colorRecycler.layoutManager =
            GridLayoutManager(
                context,
                COLOR_SPAN_COUNT,
                GridLayoutManager.HORIZONTAL,
                orientation == CameraOrientation.LANDSCAPE
            )
    }

    override fun autoFocus() {
        cameraHolder.autoFocus(viewBinding.cameraPreview)
    }

    private fun showTopMenu() = with(viewBinding) {
        if (rightMenu.visibility != View.VISIBLE) {
            rightMenu.visibility = View.VISIBLE
            rightMenu.startAnimation(inFromTopAnimation())
        }
    }

    private fun showBottomMenu() = with(viewBinding) {
        if (leftMenu.visibility != View.VISIBLE) {
            leftMenu.show()
            leftMenu.startAnimation(inFromBottomAnimation())
        }
    }

    private fun inFromTopAnimation(): Animation {
        val inFromTop = TranslateAnimation(
            Animation.RELATIVE_TO_PARENT, 0.0f,
            Animation.RELATIVE_TO_PARENT, 0.0f,
            Animation.RELATIVE_TO_PARENT, +1.0f,
            Animation.RELATIVE_TO_PARENT, 0.0f
        )
        inFromTop.duration = MENU_ANIMATION_DURATION
        inFromTop.interpolator = AccelerateInterpolator()
        return inFromTop
    }

    private fun inFromBottomAnimation(): Animation {
        val inFromBottom = TranslateAnimation(
            Animation.RELATIVE_TO_PARENT, 0.0f,
            Animation.RELATIVE_TO_PARENT, 0.0f,
            Animation.RELATIVE_TO_PARENT, -1.0f,
            Animation.RELATIVE_TO_PARENT, 0.0f
        )
        inFromBottom.duration = MENU_ANIMATION_DURATION
        inFromBottom.interpolator = AccelerateInterpolator()
        return inFromBottom
    }
}
